File: utility_billing/utility_billing/overrides/server/sales_invoice.py
import frappe
from erpnext.accounts.doctype.sales_invoice.sales_invoice import SalesInvoice
from erpnext.controllers.accounts_controller import AccountsController
from erpnext.controllers.taxes_and_totals import calculate_taxes_and_totals
from frappe import _
from frappe.model.document import Document
from frappe.utils import flt, fmt_money, getdate

from ...utils.deferred_posting import (
	cancel_deferred_journal_entries,
	create_deferred_journal_entries,
	deferred_income_accounts,
	fill_posting_dates,
	validate_deferred_lines,
)
from ...utils.item_price_scope import is_item_price_approach
from ...utils.item_rate import get_rate_for_date
from ...utils.utility_item_classifier import (
	MATCH_MULTIPLE,
	PRIMARY,
	USR_DOCTYPE,
	get_request_lines,
	get_utility_items,
	match_error_message,
	match_requests,
)
from ...utils.utils import sync_meter_readings

POSTING_DATE_FIELD = "custom_posting_date"
INVOICE_DOCTYPE = "Sales Invoice"
INVOICE_ITEM_DOCTYPE = "Sales Invoice Item"
REQUEST_FIELD = "utility_service_request"


class UtilityBillingSalesInvoice(SalesInvoice):
	"""Sales Invoice applying the deferred posting of Utility Billing.

	Deferred lines credit the deferred account of their company instead of their
	income account when the invoice is posted. The revenue reaches the income
	account through a Journal Entry dated the Deferred Posting Date of the line.
	"""

	def get_gl_entries(self, inventory_account_map=None):
		"""Build the GL entries, crediting deferred accounts for deferred lines."""
		with deferred_income_accounts(self):
			return super().get_gl_entries(inventory_account_map)


def validate(doc: Document, method: str | None = None) -> None:
	"""Validate the meter readings, the deferred posting and the utility rates of the invoice."""
	sync_meter_readings(doc)
	fill_posting_dates(doc)
	validate_deferred_lines(doc)

	if is_item_price_approach():
		govern_utility_lines(doc)


def govern_utility_lines(doc: Document) -> None:
	"""Link the invoice to its Utility Service Request and enforce its utility rates.

	Utility lines must all belong to one submitted Utility Service Request of the
	customer, which is linked when it is the only one. Primary lines are billed at
	the Item Price valid on their billing date, secondary lines at the rate on the
	request. No utility line may bill a month already billed, in this invoice or
	in a submitted one. Non utility items, lines priced from a Meter Reading and
	return invoices are left alone.

	An invoice that no longer bills any utility item has nothing left to govern,
	so a request linked earlier is cleared rather than left pointing at items the
	invoice no longer bills.

	Args:
		doc: Sales Invoice being validated.

	Raises:
		frappe.ValidationError: When the utility lines cannot be matched to one
			request, a rate does not follow its source, or a month is billed twice.
	"""
	if doc.get("is_return"):
		return

	rows = get_governed_rows(doc)
	if not rows:
		if doc.get(REQUEST_FIELD):
			doc.set(REQUEST_FIELD, None)
		return

	service_request = link_service_request(doc, [row.item_code for row in rows])
	lines = get_request_lines(service_request)
	seen_in_this_invoice: set[tuple[str, int, int]] = set()

	for row in rows:
		line = lines[row.item_code]
		billing_date = getdate(row.get(POSTING_DATE_FIELD) or doc.posting_date)
		check_not_billed_twice_in_invoice(row, billing_date, seen_in_this_invoice)

		if line.role == PRIMARY:
			apply_schedule_rate(doc, row, billing_date)
			check_not_already_invoiced(doc, row, billing_date)
		else:
			check_request_rate(doc, row, line, service_request)
			check_not_already_invoiced(doc, row, billing_date, service_request=service_request)


def get_governed_rows(doc: Document) -> list:
	"""Return the lines billing a utility item, except those priced from a Meter Reading."""
	rows = [row for row in doc.get("items") or [] if row.item_code and not row.get("meter_reading")]
	utility_items = get_utility_items(row.item_code for row in rows)

	return [row for row in rows if row.item_code in utility_items]


def link_service_request(doc: Document, item_codes: list[str]) -> str:
	"""Return the Utility Service Request governing the utility lines.

	The request is linked when it is the only one covering every utility line.

	Args:
		doc: Sales Invoice being validated.
		item_codes: Utility items billed by the invoice.

	Returns:
		The governing ``Utility Service Request`` name.

	Raises:
		frappe.ValidationError: When no single request covers every utility
			line, the linked request does not, several do and none is chosen, or
			the request is in another currency.
	"""
	match = match_requests(doc.customer, item_codes)

	message = match_error_message(match, doc.customer)
	if message:
		frappe.throw(message, title=_("Utility Service Request required"))

	service_request = doc.get(REQUEST_FIELD)

	if service_request and service_request not in match.candidates:
		frappe.throw(
			_(
				"Utility Service Request {0} does not cover every utility item on this invoice. "
				"Choose one of: {1}."
			).format(frappe.bold(service_request), ", ".join(match.candidates))
		)

	if not service_request:
		if match.status == MATCH_MULTIPLE:
			frappe.throw(
				_(
					"Several Utility Service Requests of {0} cover the utility items on this "
					"invoice. Select one in the Utility Service Request field: {1}."
				).format(frappe.bold(doc.customer), ", ".join(match.candidates))
			)

		service_request = match.candidates[0]
		doc.set(REQUEST_FIELD, service_request)

	check_request_currency(doc, service_request)

	return service_request


def check_request_currency(doc: Document, service_request: str) -> None:
	"""Block an invoice billed in another currency than its Utility Service Request."""
	currency = frappe.db.get_value(USR_DOCTYPE, service_request, "currency")

	if currency and currency != doc.currency:
		frappe.throw(
			_(
				"Utility Service Request {0} is in {1} but this invoice is in {2}. Utility "
				"lines can only be billed in the request's currency."
			).format(frappe.bold(service_request), currency, doc.currency)
		)


def check_not_billed_twice_in_invoice(row, billing_date, seen: set) -> None:
	"""Block a line billing an item for a month another line of the invoice bills."""
	key = (row.item_code, billing_date.month, billing_date.year)

	if key in seen:
		frappe.throw(
			_("Row #{0}: {1} is already billed for {2} by another row in this invoice.").format(
				row.idx, frappe.bold(row.item_code), billing_date.strftime("%B %Y")
			)
		)

	seen.add(key)


def apply_schedule_rate(doc: Document, row, billing_date) -> None:
	"""Lock a primary line to the Item Price valid on its billing date.

	Whatever the client sent, the rate is overwritten here.

	Raises:
		frappe.ValidationError: When no Item Price covers the item, the customer
			and the billing date.
	"""
	rate = get_rate_for_date(
		row.item_code,
		billing_date,
		customer=doc.customer,
		price_list=doc.selling_price_list,
		uom=row.uom,
		qty=row.qty,
		conversion_factor=row.conversion_factor or 1,
		plc_conversion_rate=doc.plc_conversion_rate or 1,
		conversion_rate=doc.conversion_rate or 1,
	)

	if rate is None:
		frappe.throw(
			_("Row #{0}: No Item Price found for {1} valid on {2}.").format(
				row.idx, frappe.bold(row.item_code), billing_date
			)
		)

	row.price_list_rate = rate
	row.rate = rate
	row.amount = flt(rate) * flt(row.qty or 1)


def check_request_rate(doc: Document, row, line, service_request: str) -> None:
	"""Block a secondary line whose rate is not the rate on its Utility Service Request.

	Raises:
		frappe.ValidationError: When the saved rate differs from the request row.
	"""
	precision = row.precision("rate")

	if flt(row.rate, precision) != flt(line.rate, precision):
		frappe.throw(
			_(
				"Row #{0}: The rate of {1} must be {2}, the rate agreed on Utility Service "
				"Request {3}."
			).format(
				row.idx,
				frappe.bold(row.item_code),
				frappe.bold(fmt_money(line.rate, currency=doc.currency)),
				frappe.bold(service_request),
			)
		)


def check_not_already_invoiced(
	doc: Document, row, billing_date, service_request: str | None = None
) -> None:
	"""Block a line billing a month already invoiced for the same item.

	Only submitted invoices count, so a draft can still be freely edited or
	discarded without permanently blocking that period. Primary items belong to
	a single property, so the item alone scopes the check. Secondary items are
	often shared by many customers, so their check is scoped to the request too.

	Args:
		doc: Sales Invoice being validated.
		row: Sales Invoice Item row being checked.
		billing_date: Resolved billing date of the row.
		service_request: Request the check is scoped to, for secondary items.

	Raises:
		frappe.ValidationError: When a submitted invoice already bills this item
			for the same month and year.
	"""
	invoice = frappe.qb.DocType(INVOICE_DOCTYPE)
	item = frappe.qb.DocType(INVOICE_ITEM_DOCTYPE)

	query = (
		frappe.qb.from_(item)
		.join(invoice)
		.on(item.parent == invoice.name)
		.select(invoice.name)
		.where(invoice.docstatus == 1)
		.where(invoice.name != (doc.name or ""))
		.where(item.item_code == row.item_code)
		.where(frappe.qb.terms.Function("MONTH", item[POSTING_DATE_FIELD]) == billing_date.month)
		.where(frappe.qb.terms.Function("YEAR", item[POSTING_DATE_FIELD]) == billing_date.year)
		.limit(1)
	)

	if service_request:
		query = query.where(invoice[REQUEST_FIELD] == service_request)

	existing = query.run()
	if existing:
		frappe.throw(
			_("Row #{0}: {1} was already invoiced for {2} on {3}.").format(
				row.idx,
				frappe.bold(row.item_code),
				billing_date.strftime("%B %Y"),
				frappe.bold(existing[0][0]),
			)
		)


def before_validate(doc: Document, method: str) -> None:
	"""Intercepts submit event for document"""
	if not doc.taxes:
		AccountsController.append_taxes_from_item_tax_template(doc)
		calculate_taxes_and_totals(doc)
	unique_sales_orders = {
		item.sales_order
		for item in frappe.get_all(
			"Sales Invoice Item",
			filters={"parent": doc.name, "sales_order": ["is", "set"]},
			fields=["sales_order"],
		)
	}

	for sales_order in unique_sales_orders:
		map_sales_order_meter_readings_to_invoice(sales_order, doc)


def map_sales_order_meter_readings_to_invoice(sales_order_name, target_doc):
	"""Map all fields from Sales Order Meter Reading to Sales Invoice Meter Reading."""
	sales_order = frappe.get_doc("Sales Order", sales_order_name)
	meter_readings = sales_order.get("meter_readings")
	target_doc.set("meter_readings", [])
	if sales_order.utility_property:
		target_doc.utility_property = sales_order.utility_property
	if sales_order.utility_service_request:
		target_doc.utility_service_request = sales_order.utility_service_request

	for reading in meter_readings:
		new_reading_data = reading.as_dict()
		new_reading_data.pop("name", None)
		new_reading = target_doc.append("meter_readings", new_reading_data)
		new_reading.parent = target_doc.name


def on_submit(doc: Document, method: str) -> None:
	"""Recognise the deferred revenue of the invoice on its deferred dates."""
	create_deferred_journal_entries(doc)


def on_cancel(doc: Document, method: str) -> None:
	"""Cancel the journal entries that recognised the deferred revenue."""
	cancel_deferred_journal_entries(doc)

File: utility_billing/utility_billing/overrides/client/sales_invoice.js
// Line helpers of the Sales Invoice item grid.
//
// A line whose revenue is deferred to a later date has to be billed at the rate
// valid on that date, and a rent line bills a property through the service item
// of that property: both are kept in step here.
//
// Under the Item Price billing approach, utility lines (items flagged
// `is_utility_item`) are also governed by a Utility Service Request: the invoice
// is linked to the one request of the customer covering all of them, that
// request's utility items are pre-filled, and their rates follow the request -
// primary lines the Item Price valid on their date, secondary lines the rate on
// the request row. A rate typed by hand is put back when the field is left. The
// server enforces all of this again on save; this is UX only.

// Rates already resolved per line (line, item and date), so a form refresh or a
// rate edited by hand does not resolve the same rate over and over. The var form
// keeps the cache when the script is evaluated again on the same page.
var resolved_line_rates = resolved_line_rates || {};

var UTILITY_API = "utility_billing.api.sales_invoice";

frappe.ui.form.on("Sales Invoice", {
	setup: function (frm) {
		set_service_request_query(frm);
	},

	// The posting dates of the lines are filled on the server when the invoice is
	// saved, so their rates are brought in line with them when the form opens.
	refresh: function (frm) {
		(frm.doc.items || []).forEach((row) => {
			apply_rate_of_deferred_date(frm, row.doctype, row.name);
		});

		load_service_request_lines(frm, false);
	},

	customer: function (frm) {
		govern_utility_lines(frm);
	},

	utility_service_request: function (frm) {
		load_service_request_lines(frm, true);
	},
});

frappe.ui.form.on("Sales Invoice Item", {
	item_code: function (frm, cdt, cdn) {
		set_property_of_item(frm, cdt, cdn);
		govern_utility_lines(frm, cdn);
	},

	utility_property: function (frm, cdt, cdn) {
		set_item_of_property(frm, cdt, cdn);
	},

	items_remove: function (frm) {
		govern_utility_lines(frm);
	},

	custom_posting_date: function (frm, cdt, cdn) {
		if (enforce_governed_rate(frm, cdt, cdn)) return;
		apply_rate_of_deferred_date(frm, cdt, cdn, true);
	},

	// ERPNext applies the rate of the invoice posting date to these fields, so
	// both are used to correct the rate to the deferred posting date afterwards.
	price_list_rate: function (frm, cdt, cdn) {
		if (revert_governed_rate(frm, cdt, cdn, "price_list_rate")) return;
		apply_rate_of_deferred_date(frm, cdt, cdn);
	},

	rate: function (frm, cdt, cdn) {
		if (revert_governed_rate(frm, cdt, cdn, "rate")) return;
		apply_rate_of_deferred_date(frm, cdt, cdn);
	},
});

// Resolve whether rent is billed by Item Price; governance applies only then.
function is_item_price_approach() {
	return frappe.db
		.get_single_value("Utility Billing Settings", "rent_billing_approach")
		.then((approach) => approach === "Item Price");
}

// Only submitted requests of the customer can be linked, narrowed to the
// requests covering every utility line once they are known.
function set_service_request_query(frm) {
	frm.set_query("utility_service_request", () => {
		const filters = { docstatus: 1 };

		if (frm.doc.customer) {
			filters.customer = frm.doc.customer;
		}
		if ((frm.utility_request_candidates || []).length) {
			filters.name = ["in", frm.utility_request_candidates];
		}

		return { filters };
	});
}

// Match the utility lines of the invoice to the customer's requests, once both
// a utility item and the customer are set - whichever comes first.
function govern_utility_lines(frm, cdn) {
	if (frm.doc.docstatus !== 0 || frm.doc.is_return || frm.prefilling_utility_lines) {
		return;
	}

	const item_codes = [
		...new Set(
			(frm.doc.items || [])
				.filter((row) => row.item_code && !row.meter_reading)
				.map((row) => row.item_code),
		),
	];

	if (!frm.doc.customer || !item_codes.length) {
		frm.utility_request_candidates = [];
		clear_orphaned_service_request(frm);
		return;
	}

	const new_row = cdn ? locals["Sales Invoice Item"][cdn] : null;
	const sequence = (frm.utility_match_sequence = (frm.utility_match_sequence || 0) + 1);

	is_item_price_approach().then((applies) => {
		if (!applies) return;

		frappe.call({
			method: `${UTILITY_API}.match_utility_service_requests`,
			args: {
				customer: frm.doc.customer,
				item_codes: item_codes,
				new_item_code: new_row ? new_row.item_code : null,
			},
			callback: function (r) {
				if (sequence !== frm.utility_match_sequence) return; // superseded

				// Picking an item makes ERPNext send the invoice to the server
				// (process_item_selection) and sync the server copy back when it
				// answers, deleting any field set on the form in between. Acting
				// once no request is in flight keeps the link and rates in place.
				frappe.after_ajax(() => {
					if (sequence !== frm.utility_match_sequence) return; // superseded meanwhile

					handle_request_match(frm, r.message || {}, cdn);
				});
			},
		});
	});
}

// Clear a linked request once the invoice no longer bills any of its utility
// items - an invoice with no utility item has nothing left to govern.
function clear_orphaned_service_request(frm) {
	if (frm.doc.utility_service_request) {
		frm.set_value("utility_service_request", null);
	}
}

function handle_request_match(frm, match, cdn) {
	frm.utility_request_candidates = match.applies ? match.candidates || [] : [];

	if (!match.applies) {
		clear_orphaned_service_request(frm);
		return;
	}

	if (!match.status) {
		// No utility item on the invoice at all.
		clear_orphaned_service_request(frm);
		return;
	}

	if (match.status === "missing" || match.status === "conflict") {
		block_utility_line(frm, match, cdn);
		return;
	}

	const linked = frm.doc.utility_service_request;

	if (linked && match.candidates.includes(linked)) {
		if (cdn) enforce_governed_rate(frm, "Sales Invoice Item", cdn);
		return;
	}

	if (match.status === "single") {
		frm.set_value("utility_service_request", match.candidates[0]);
		return;
	}

	// Several requests qualify: the user picks one, the picker lists only them.
	if (linked) {
		frm.set_value("utility_service_request", null);
	}

	frappe.show_alert(
		{
			message: __(
				"Several Utility Service Requests of {0} cover the utility items on this invoice. Select one in the Utility Service Request field.",
				[frm.doc.customer],
			),
			indicator: "orange",
		},
		10,
	);
	frm.scroll_to_field("utility_service_request");
}

// Remove the line that cannot be billed. When the customer changed instead, the
// lines stay for the user to fix and the request is unlinked.
function block_utility_line(frm, match, cdn) {
	frappe.msgprint({
		title: __("Utility item not allowed"),
		message: match.message,
		indicator: "red",
	});

	const row = cdn ? locals["Sales Invoice Item"][cdn] : null;
	const is_culprit =
		row && (match.status === "conflict" || (match.missing || []).includes(row.item_code));
	const grid_row = is_culprit ? frm.fields_dict.items.grid.get_row(cdn) : null;

	if (grid_row) {
		grid_row.remove();
		return;
	}

	if (frm.doc.utility_service_request) {
		frm.set_value("utility_service_request", null);
	}
}

// Load the governed lines of the linked request and bring every line in step.
// When the request was just linked its utility items are pre-filled first.
function load_service_request_lines(frm, prefill) {
	frm.utility_lines = {};
	frm.utility_expected_rates = {};

	const request = frm.doc.utility_service_request;

	if (frm.doc.docstatus !== 0 || frm.doc.is_return || !request) {
		return;
	}

	is_item_price_approach().then((applies) => {
		if (!applies) return;

		frappe.call({
			method: `${UTILITY_API}.get_utility_service_request_lines`,
			args: { utility_service_request: request },
			callback: function (r) {
				if (frm.doc.utility_service_request !== request) return; // relinked meanwhile

				const lines = r.message || [];
				lines.forEach((line) => {
					frm.utility_lines[line.item_code] = line;
				});

				const ready = prefill ? prefill_service_request_lines(frm, lines) : Promise.resolve();

				// Pre-filled items are still being processed by ERPNext; their
				// answers would overwrite the rates if these were set first.
				ready
					.then(() => frappe.after_ajax())
					.then(() => {
						if (frm.doc.utility_service_request !== request) return; // relinked meanwhile

						(frm.doc.items || []).forEach((row) => {
							enforce_governed_rate(frm, row.doctype, row.name);
						});
					});
			},
		});
	});
}

// Add the request's utility items the invoice does not bill yet.
async function prefill_service_request_lines(frm, lines) {
	const billed = new Set((frm.doc.items || []).map((row) => row.item_code));
	const missing = lines.filter((line) => !billed.has(line.item_code));

	if (!missing.length) return;

	frm.prefilling_utility_lines = true;

	try {
		for (const line of missing) {
			const row = (frm.doc.items || []).find((item) => !item.item_code) || frm.add_child("items");

			await frappe.model.set_value(row.doctype, row.name, "item_code", line.item_code);
			await frappe.model.set_value(row.doctype, row.name, "qty", line.qty);
		}
	} finally {
		frm.prefilling_utility_lines = false;
	}

	frm.refresh_field("items");
}

// The governed line of the linked request billed by a row, if any.
function get_governed_line(frm, row) {
	if (!row || !row.item_code || row.meter_reading) return null;
	if (frm.doc.docstatus !== 0 || frm.doc.is_return) return null;

	return (frm.utility_lines || {})[row.item_code] || null;
}

// Set the rate a governed row must bill. Returns whether the row is governed.
function enforce_governed_rate(frm, cdt, cdn) {
	const row = locals[cdt][cdn];
	const line = get_governed_line(frm, row);

	if (!line) return false;

	if (line.role === "secondary") {
		set_governed_rate(frm, row, flt(line.rate));
		return true;
	}

	frappe.call({
		method: `${UTILITY_API}.get_line_rate`,
		args: {
			item_code: row.item_code,
			posting_date: row.custom_posting_date || frm.doc.posting_date,
			customer: frm.doc.customer,
			price_list: frm.doc.selling_price_list,
			uom: row.uom,
			qty: row.qty,
			conversion_factor: row.conversion_factor,
			plc_conversion_rate: frm.doc.plc_conversion_rate,
			conversion_rate: frm.doc.conversion_rate,
		},
		callback: function (r) {
			// A missing Item Price is reported by the server on save.
			if (r.message === null || r.message === undefined) return;
			if (!locals[cdt][cdn] || locals[cdt][cdn].item_code !== line.item_code) return;

			set_governed_rate(frm, locals[cdt][cdn], flt(r.message));
		},
	});

	return true;
}

function set_governed_rate(frm, row, rate) {
	frm.utility_expected_rates = frm.utility_expected_rates || {};
	frm.utility_expected_rates[row.name] = rate;

	const values = {};

	if (flt(row.price_list_rate) !== rate) values.price_list_rate = rate;
	if (flt(row.rate) !== rate) values.rate = rate;

	if (Object.keys(values).length) {
		frappe.model.set_value(row.doctype, row.name, values);
	}
}

// Put back a governed rate edited by hand. Returns whether the row is governed.
function revert_governed_rate(frm, cdt, cdn, fieldname) {
	const row = locals[cdt][cdn];

	if (!get_governed_line(frm, row)) return false;

	const expected = (frm.utility_expected_rates || {})[cdn];

	if (expected === undefined) return true; // still resolving

	if (flt(row[fieldname], precision(fieldname, row)) === flt(expected, precision(fieldname, row))) {
		return true;
	}

	frappe.model.set_value(cdt, cdn, fieldname, expected);
	frappe.show_alert({
		message: __("The rate of {0} is set by Utility Service Request {1} and cannot be changed.", [
			row.item_code,
			frm.doc.utility_service_request,
		]),
		indicator: "orange",
	});

	return true;
}

// Fill the property billed by the selected item, when the item is the service
// item of a property.
function set_property_of_item(frm, cdt, cdn) {
	const row = locals[cdt][cdn];

	if (frm.doc.docstatus !== 0 || !row.item_code) {
		return;
	}

	frappe.call({
		method: "utility_billing.api.sales_invoice.get_line_utility_property",
		args: { item_code: row.item_code },
		callback: function (r) {
			const property = r.message;

			if (property && property !== row.utility_property) {
				frappe.model.set_value(cdt, cdn, "utility_property", property);
			}
		},
	});
}

// Fill the service item of the selected property, when the line has no item yet.
function set_item_of_property(frm, cdt, cdn) {
	const row = locals[cdt][cdn];

	if (frm.doc.docstatus !== 0 || !row.utility_property || row.item_code) {
		return;
	}

	frappe.call({
		method: "utility_billing.api.sales_invoice.get_line_item",
		args: { utility_property: row.utility_property },
		callback: function (r) {
			if (r.message) {
				frappe.model.set_value(cdt, cdn, "item_code", r.message);
			}
		},
	});
}

// Replace the rate of the line by the rate valid on its Deferred Posting Date.
// Governed utility lines are kept in step by their request instead.
//
// The rate resolved for a line is remembered by line, item and date, so a form
// refresh or a rate edited by hand does not resolve it over and over.
function apply_rate_of_deferred_date(frm, cdt, cdn, force) {
	const row = locals[cdt][cdn];

	if (frm.doc.docstatus !== 0 || frm.doc.is_return) {
		return;
	}

	if (!row.item_code || !row.custom_posting_date || get_governed_line(frm, row)) {
		return;
	}

	if (row.custom_posting_date === frm.doc.posting_date) {
		return; // the rate of the invoice posting date is already applied
	}

	const key = `${row.name}|${row.item_code}|${row.custom_posting_date}`;

	if (!force && resolved_line_rates[key] === flt(row.price_list_rate)) {
		return; // the rate of the deferred posting date is already applied
	}

	frappe.call({
		method: "utility_billing.api.sales_invoice.get_line_rate",
		args: {
			item_code: row.item_code,
			posting_date: row.custom_posting_date,
			customer: frm.doc.customer,
			price_list: frm.doc.selling_price_list,
			uom: row.uom,
			qty: row.qty,
			conversion_factor: row.conversion_factor,
			plc_conversion_rate: frm.doc.plc_conversion_rate,
			conversion_rate: frm.doc.conversion_rate,
		},
		callback: function (r) {
			const rate = r.message;

			if (rate === null || rate === undefined) {
				return;
			}

			resolved_line_rates[key] = flt(rate);

			if (flt(row.price_list_rate) !== flt(rate)) {
				// ERPNext derives the rate and the totals from the price list rate.
				frappe.model.set_value(cdt, cdn, "price_list_rate", rate);
			}
		},
	});
}